dsc_lib/src/parser/mod.rs

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use expressions::Expression;
+use tree_sitter::Parser;
+
+use crate::dscerror::DscError;
+use crate::functions::FunctionDispatcher;
+
+pub mod expressions;
+pub mod functions;
+
+pub struct Statement {
+    parser: Parser,
+    function_dispatcher: FunctionDispatcher,
+}
+
+impl Statement {
+    /// Create a new `StatementParser` instance.
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if the underlying parser fails to initialize.
+    pub fn new() -> Result<Self, DscError> {
+        let mut parser = Parser::new();
+        parser.set_language(tree_sitter_dscexpression::language())?;
+        let function_dispatcher = FunctionDispatcher::new();
+        Ok(Self {
+            parser,
+            function_dispatcher,
+        })
+    }
+
+    /// Parse and execute a statement.
+    ///
+    /// # Arguments
+    ///
+    /// * `statement` - The statement to parse and execute.
+    ///
+    /// # Errors
+    ///
+    /// This function will return an error if the statement fails to parse or execute.
+    pub fn parse_and_execute(&mut self, statement: &str) -> Result<String, DscError> {
+        let Some(tree) = &mut self.parser.parse(statement, None) else {
+            return Err(DscError::Parser("Error parsing statement".to_string()));
+        };
+        let root_node = tree.root_node();
+        if root_node.is_error() {
+            return Err(DscError::Parser("Error parsing statement root".to_string()));
+        }
+        let root_node_kind = root_node.kind();
+        if root_node_kind != "statement" {
+            return Err(DscError::Parser("Invalid statement".to_string()));
+        }
+        let Some(child_node) = root_node.named_child(0) else {
+            return Err(DscError::Parser("Child node not found".to_string()));
+        };
+        if child_node.is_error() {
+            return Err(DscError::Parser("Error parsing statement child".to_string()));
+        }
+        let kind = child_node.kind();
+        match kind {
+            "stringLiteral" | "bracketInStringLiteral" => {
+                let Ok(value) = child_node.utf8_text(statement.as_bytes()) else {
```

Comment:
`statement.as_bytes()` is called many times when traverse the tree, which is unneeded overhead. I think you should call it once in this method, and use the returned value for the rest places.

Reply:
`as_bytes()` returns a slice, so seems ok here vs allocating a variable that might not get used.

Reply:
It should always be used when reaching the line `match kind`.

`as_bytes` converts a string slice into a byte slice through `mem::transmute<Src, Dst>(src: Src) -> Dst`, whose documentation indicates a copy operation. Also, in general, avoiding calling a method multiple times to get the same value is a good practice to follow.

> `transmute` is semantically equivalent to a bitwise move of one type
> into another. It copies the bits from the source value into the
> destination value, then forgets the original. Note that source and destination
> are passed by-value, which means if `Src` or `Dst` contain padding, that padding
> is *not* guaranteed to be preserved by `transmute`.

Reply:
Didn't realize it did a copy.  Will change.